### PR TITLE
Add DisableGarbageCollect annotation

### PR DESCRIPTION
**Pull request by phillebaba**
This change implements an annotation to disable garbage collection for specific resources.

https://github.com/fluxcd/flux/issues/2841

### CODE REVIEW

File: pkg/cluster/kubernetes/sync_test.go [hunk 1]
```diff
@@ -464,6 +485,18 @@ metadata:
 		test(t, kube, "", ns1+defs1+defs2+ns3+defs3, false)
 	})
 
+	t.Run("sync adds and GCs skips resources", func(t *testing.T) {
```
Comment (squaremo) [inline pkg/cluster/kubernetes/sync_test.go]: This test does verify what it claims to; but I think it should also verify that resources are updated (i.e., not ignored) when they have a gc_disabled annotation.

File: pkg/cluster/kubernetes/sync.go [hunk 2]
```diff
@@ -145,6 +145,11 @@ func (c *Cluster) collectGarbage(
 
 		switch {
 		case !ok: // was not recorded as having been staged for application
+			if res.Policies().Has(policy.DisableGarbageCollect) {
+				c.logger.Log("info", "not deleting resource; garbage colelction annotation in file", "resource", res.ResourceID())
```
Comment (squaremo) [inline pkg/cluster/kubernetes/sync.go]:
```suggestion
				c.logger.Log("info", "not deleting resource; garbage collection disabled by annotation", "resource", res.ResourceID())
```

(typo, and the annotation is in the cluster resource, not the file)

**Comment by squaremo** [comment]
Thinking aloud: I wonder if this could be value of "ignore", instead of its own annotation.

Here are the combinations of annotations and effects, when there are both:

| ignore \ disable_gc | false | true |
|---------------------------|--------|--------|
| **false**                  | sync and gc | sync, no gc |
| **true**                    | no sync, no gc | no sync, no gc |

There's only three different behaviours there, so they don't operate independently -- one of them is redundant. So you could represent those three possibilities as three values:

    fluxcd.io/ignore: "true" | "false" | "gc"

One counter-argument is that it's weird to have a three-valued boolean like that (albeit they are only stringly pictures of booleans).

wdyt @phillebaba ?

**Comment by phillebaba** [comment]
There would be some benefit with reducing the number of annotations, the risk is that having three values would confuse people. So documentation of the behavior is important.

The `ingore: true` + `disable_gc: false` combination would only disable gc if we assume ignore works as it was intended from the beginning.

Does these descriptions match the intended behavior of the different values? Assuming that the end user has enabled GC.
| value | description | 
| --- | --- | 
| false | Resources are synced when added to git and garbage collected when removed from git |
| no_gc | Resources are synced when added to git but will not be garbage collected when removed from git. |
| true | Resources are not synced when added to git. Additionally if the value is set manually outside of flux the resource will no longer be synced by flux, and it will not be garbage collected if removed from git. |

Am I missing something in the description?

**Comment by squaremo** [comment]
> There would be some benefit with reducing the number of annotations, the risk is that having three values would confuse people. So documentation of the behavior is important.

Yes, I agree on both points. There's an opportunity to change the formulation slightly, so long as it's backward-compatible -- say, `ignore: {false,none}|sync_only|{true,sync_and_gc}`. If that would mitigate the confusion ..

> The `ingore: true` + `disable_gc: false` combination would only disable gc if we assume ignore works as it was intended from the beginning.

Right, that's presupposed. I find it hard to imagine that someone is relying on it _not_ working like that (even I had difficulty getting my head around the fact that it didn't).

> Resources are synced when added to git but will not be garbage collected when removed from git.

I would say "synced _from_ git", since updates in git will still be applied to the cluster. Similarly, in the description of "true".

**Comment by squaremo** [comment]
OK so I think the goal for this PR should be to make it work as we've settled on above -- are you happy to adjust it, @phillebaba? Alternatively, I could make ignore mean ignore-for-gc-too in another PR, then you could work on top of that.

**Comment by phillebaba** [comment]
I can implement both parts, don't think that is an issue. You are right about going with the multi value annotation solution as it would simplify logic.

I have a lot to do at work this week so I can't get this done until the weekend.

**Comment by squaremo** [comment]
Cool, no worries @phillebaba, any time you lend to this project is appreciated :-)

**Comment by phillebaba** [comment]
I change the behavior so that you disable GC by setting `sync_only` as the value for the ignore annotation.

I also added some more info in the FAQ, but i think that getting a single page that documents all annotations as described in https://github.com/fluxcd/flux/issues/2671 would be great!

A table like this would describe the ignore annotations behavior pretty well.
| Value | Sync | GC |
| --- | --- | --- |
| false | x | x |
| sync_only | x |  |
| true |  |  |

File: docs/faq.md [hunk 3]
```diff
@@ -341,6 +341,18 @@ how/where to undo the change (cf [flux#1211](https://github.com/fluxcd/flux/issu
 annotation exists in either the cluster or in git, it will be respected, so you may need to remove
 it from both places.
 
+Additionally when garbage collection is enabled. Flux will not garbage collect resources in the cluster
```
Comment (squaremo) [inline docs/faq.md]:
```suggestion
Additionally, when garbage collection is enabled, Flux will not garbage collect resources in the cluster
```

File: docs/faq.md [hunk 4]
```diff
@@ -341,6 +341,18 @@ how/where to undo the change (cf [flux#1211](https://github.com/fluxcd/flux/issu
...
+with the ignore annotation if the resource is removed from git.
+
+### Can I disable garbage collection for a specific resource?
+
+Yes. By adding the annotation below to a resource Flux will sync updates from git, but it will not
+garbage collect when the resource is removed from git.
```
Comment (squaremo) [inline docs/faq.md]: :+1:

File: pkg/cluster/kubernetes/sync.go [hunk 5]
```diff
@@ -145,6 +145,17 @@ func (c *Cluster) collectGarbage(
 
 		switch {
 		case !ok: // was not recorded as having been staged for application
+			if res.Policies().Has(policy.Ignore) {
+				c.logger.Log("info", "skipping GC of cluster resource; resource has ignore policy true", "dry-run", dryRun, "resource", resourceID)
+				continue
+			}
+
+			v, ok := res.Policies().Get(policy.Ignore)
+			if ok && v == "sync_only" {
```
Comment (squaremo) [inline pkg/cluster/kubernetes/sync.go]: Could do with a constant for this

File: pkg/cluster/kubernetes/sync_test.go [hunk 6]
```diff
@@ -703,7 +704,74 @@ spec:
...
+  name: dep1
+  namespace: foobar
+  annotations: {flux.weave.works/ignore: "sync_only"}
+`
+
+		// sync namespace and deployment
+		test(t, kube, ns1+dep1, ns1+dep1, false)
+
+		// sync namespace only but expect deployment to not be GC
+		test(t, kube, ns1, ns1+dep1, false)
```
Comment (squaremo) [inline pkg/cluster/kubernetes/sync_test.go]: Nice and simple test case :+1:

**Pull request merged**